### DISCUSSION TITLE
Feature requests change TTS generator file names and a dictionary function

### DISCUSSION THREAD

**Discussion by ghost** [erew123/alltalk_tts#214] [Q&A]
It would be  nice if there was an option to automatically name the output files like the names in the List e.g. ID1, or even better : 
ID 001.wav.

It would be good to add a dictonary function so that yo can set one word to be spoken as if it was another.

**Comment by erew123** [comment]
Hi @Larrycmd 

There is an option to do that through the API, however, because of the way certain things work, you have to add an element of randomness to the names e.g. Lets say you are in text-gen-webui and you are chatting in there with a LLM bot. If you specify the name as name_001, name_002 etc then first of you have to read back the files list to always track where you are in the numbering. Over time this will slow things down as you have to read out the whole folder contents. Additionally if you dont track the names, you may close a session and re-open a session later on, which would result in the the files being over-written if the current numbering isn't tracked and even then, there's a further complication. Lets say you use that naming scheme mentioned above, you have generated 100 files name_001 to name_100, but you delete the first 50 files as you want to save name_050 to name_100, you then have a complication tracking the fact that its ok to use name_001 to name_49, but not 50 to 100 etc..

Im just pointing out it can get a bit complicated, hence there is always a need to add a unique ID number to the files in some way, otherwise you can always end up with the potential to over-write previous TTS generations.

So changing the "name" portion of the file is easy, but naming the unique ID portion of the file can be complicated. 

Is there a specific section of AllTalk you want to target sequential numbering at? text-gen-webui, tts-generator, the API etc?

Thanks

**Reply by ghost** [reply]
Yes, TTS Generator, because if there are errors in one  sentence it would be easier to memorize the wrong one, generate the rest and than regenerate the wrong one.

**Comment by greek12man** [comment]
I would also like this. How about using a naming system similar to what 11labs does?
Starting with the date so its always unique
2024-05-10T11_52_28_

**Reply by erew123** [reply]
This can be difficult. In an earlier version of AllTalk I did encounter a situation where there is a race situation if 2x generations occur at the same time, whereby if the generation request happens in the same second (because we are making multiple generations) the timestamp ended up the same, hence one file was wiped/overwritten by the first. So there is always an argument to a through a uuid (unique identifier) somewhere into the file name.

**Comment by erew123** [comment]
@Larrycmd @greek12man So tell me if I am getting this correct, what you are both asking for is an easier way to identify the visual of what you see on screen in the TTS_generator (line number or whatever) to match back to the file name somehow? Is that a fair summary?

**Reply by erew123** [reply]
@Larrycmd @greek12man and perhaps to name the files with a unique name, linked to the current TTS generation you are doing e.g. if you are generating a story called "My Story" you would like the files to be named "My Story"

**Reply by greek12man** [reply]
I usually mass generate:
Line1
line2
line3

and want the audio files to be something like, audio001, audio002, audio003

I tend to solve this by sorting by date, but i encounter one problem sometimes. line 3 sometimes generated before line 2.
I suspect this is because line 3 is very short and line 2 is very long.

From what you described, this wouldnt solve this problem, from what i undestand.

For my personal need, what would be even better, would be to feed the program a csv file.
One column my lines, one column the file names i want each line audio generated, to be renamed to. Probably too difficult.

**Reply by ghost** [reply]
@erew123 
the nummeric value would be even more important to me than a project name( in your Example (my story), for your example it would be 001mystory_uniquevalue.wav
and for  regenerateted files it would be nice if they could keep a name simlar to the orignal file lets say 001mystory__re1uniquevalue.wav or somthing like that for the first regeneration  of a file and so on. Awsome programm by the way thanks for this work.

**Comment by q5sys** [comment]
> @Larrycmd @greek12man and perhaps to name the files with a unique name, linked to the current TTS generation you are doing e.g. if you are generating a story called "My Story" you would like the files to be named "My Story"

What about keeping it mostly the way it is now, but including a function that appends a user provided prefix to the file name? That would allow people to use either a name or a date that they would manually enter.
So instead of `TTS_17153280395f546` being the file name, it'd be `mystory_TTS_17153280395f546` and if they want date stamps, they can add that into the user definable prefix field... `2024-05-10_mystory_TTS_17153280395f546`

**Comment by erew123** [comment]
Without going too in-depth on the inner working of things, I've put some code up for you to try, After you have performed a `git pull` to get the latest version down, you will be able to use the **test** version of the tts generator on:

http://127.0.0.1:7851/static/tts_generator/tts_generator_test.html

You will manually have to go to that link! (Note the `_test` bit added to the URL/name). You should see this:

![image](https://github.com/erew123/alltalk_tts/assets/35898566/1829c740-bb27-48d2-a192-0e742b5be950)

Its been a bit of an ass to rework, but my testing seems to show it works, though youre all welcome to test.

![image](https://github.com/erew123/alltalk_tts/assets/35898566/6d942f8c-f9c1-4a0c-a02c-5ebfd5c319c4)

The format of the files is:

**{Custom-file-name}** _ **{ID number up to 99,999}** _ **{Unique_ID_number}**

There is simply no way to avoid having a unique ID number, both on generation and re-generation, without further web-browser caching issues.

Im trying to spend most of my time coding AllTalk v2, so you are welcome to give this a go and feed back, however, as mentioned I may be moving over to Gradio for this, so wont be focussing too strongly on more code updates to this version.

Thanks

**Reply by greek12man** [reply]
It seems to work great so far, thank you. I will write back if I find any issue.
